### PR TITLE
Cosmic Reach Model Editor plugin

### PR DESCRIPTION
**Pull request by zHoeshin**
Add the Cosmic Reach Model Editor plugin to the list of Blockbench plugins

### CODE REVIEW

File: plugins.json [hunk 1]
```diff
@@ -889,5 +889,17 @@
...
+		"title": "Cosmic Reach Model Editor",
+		"author": "Z. Hoeshin",
+		"icon": "icon.png",
+		"description": "Allows creating, editing, importing and exporting Cosmic Reach block models.",
+		"tags": ["Cosmic Reach"],
+		"version": "1.1.0",
+		"min_version": "4.9.4",
+		"variant": "both",
+		"creation_date": "2024-03-09",
+		"new_repository_format": true,
```
Comment (Malik12tree) [inline plugins.json:903]: The plugin's meta data is mismatched with the meta data defined in the code. (mismatched `tags`, missing `min_version`, `variant`, `creation_date`, `new_repository_format`)
```js
    Plugin.register(id, {
      title: name,
      icon: "icon.png",
      author: "Z. Hoeshin",
      description: "Allows creating, editing, importing and exporting Cosmic Reach block models.",
      tags: [],
      version: "1.1.0",
      creation_date: "2024-03-09",
```

File: plugins/cosmic_reach_model_editor/cosmic_reach_model_editor.js [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+      icon: "icon.png",
+      author: "Z. Hoeshin",
+      description: "Allows creating, editing, importing and exporting Cosmic Reach block models.",
+      tags: [],
+      version: "1.1.0",
+      creation_date: "2024-03-09",
+      onload() {
+        Codecs.java_block.load_filter.condition = (model) => {
+			return (model.parent || model.elements || model.textures) && (!model.cuboids);
+		}
```
Comment (Malik12tree) [inline plugins/cosmic_reach_model_editor/cosmic_reach_model_editor.js:26]: You should not directly overwrite the condition for the load filter, as its internal condition is subject to change whether inside Blockbench or an another plugin. You should store the original condition and make use of it inside the overwritten condition.

Define a reference in the plugin's scope.
```js
let originalJavaBlockLoadFilter;
```
```suggestion
        originalJavaBlockLoadFilter = Codecs.java_block.load_filter.condition;
        Codecs.java_block.load_filter.condition = (model) => {
		return originalJavaBlockLoadFilter(model)  && (!model.cuboids);
	}
```
And do not forget onunload to reset the condition..
```js
Codecs.java_block.load_filter.condition = originalJavaBlockLoadFilter;
```

File: plugins/cosmic_reach_model_editor/cosmic_reach_model_editor.js [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+function sumArrays(a, b){
+    s = []
+    for(let i = 0; i < Math.min(a.length, b.length); i++){
+        s.push(a[i] + b[i])
+    }
+    return s
+}
+
+(() => {
+    let codec, format, export_action, import_action, dialog, properties
```
Comment (Malik12tree) [inline plugins/cosmic_reach_model_editor/cosmic_reach_model_editor.js:11]: `sumArray` should be defined in the plugin's local scope and not exposed globally.
```suggestion
(() => {
    function sumArrays(a, b){
        s = []
        for(let i = 0; i < Math.min(a.length, b.length); i++){
            s.push(a[i] + b[i])
        }
        return s
    }
    let codec, format, export_action, import_action, dialog, properties
```

**Review by Malik12tree — CHANGES_REQUESTED** [review]

**Comment by Malik12tree** [comment]
Please make use of `originalJavaBlockCond`, To clarify again, the original condition might be changed within Blockbench or by other plugins. Instead, use the saved condition within your new condition. This will protect your plugin from breaking Blockbench or other plugins in the future.

**Comment by Malik12tree** [comment]
Yeah of course! But only to give your plugin more availability in case the original condition (`model.parent || model.elements || model.textures`) changes, instead of hard coding it in your code, just change it to call the original condition. It will behave exactly as it's behaving now, only less subject to break things in the future.

```js
(mode) => {
    return !mode.cuboids && originalJavaBlockCond(model);
}
```

**Comment by zHoeshin** [comment]
@Malik12tree , the issue's long ago fixed

**Comment by Malik12tree** [comment]
@zHoeshin Yeah totally aware. Jannis is the only person that can merge pull requests in this repo, sometimes pull requests may take long if he's busy

**Comment by JannisX11** [comment]
Thanks for your help @Malik12tree !

There are two breaking issure remaining:
* `Codec.java_block` does not exist. Use `Codecs.java_block` instead in the 2 places where it's misspelled
* plugins.json has a trailing comma, as you can seen from the automatic check above

I would also strongly recommend to add icons to your model format and your actions.
Once this is fixed, you may update the creation date field to the current date so it'll show up as new.

**Comment by JannisX11** [comment]
The new icons are not working correctly.
It's a good practice to always test all the changes that you make.
image

**Pull request merged**